File: 2025-07-July-GFG-POTD/July 01 - Substrings of length k with k-1 distinct elements/Solution.py
from collections import defaultdict

class Solution:
    def substrCount(self, s: str, k: int) -> int:
        if len(s) < k:
            return 0

        freq = defaultdict(int)
        count = 0

        # First window
        for i in range(k):
            freq[s[i]] += 1
        if len(freq) == k - 1:
            count += 1

        # Slide window
        for i in range(k, len(s)):
            out_char = s[i - k]
            freq[out_char] -= 1
            if freq[out_char] == 0:
                del freq[out_char]

            in_char = s[i]
            freq[in_char] += 1

            if len(freq) == k - 1:
                count += 1

        return count

File: 2025-07-July-GFG-POTD/July 01 - Substrings of length k with k-1 distinct elements/Solution.cpp
#include <unordered_map>
#include <string>
using namespace std;

class Solution {
public:
    int substrCount(string s, int k) {
        if (s.size() < k) return 0;

        unordered_map<char, int> freqMap;
        int count = 0;

        // Initialize first window
        for (int i = 0; i < k; ++i) {
            freqMap[s[i]]++;
        }
        if (freqMap.size() == k - 1) {
            count++;
        }

        // Slide the window
        for (int i = k; i < s.size(); ++i) {
            char outChar = s[i - k];
            freqMap[outChar]--;
            if (freqMap[outChar] == 0) {
                freqMap.erase(outChar);
            }

            char inChar = s[i];
            freqMap[inChar]++;

            if (freqMap.size() == k - 1) {
                count++;
            }
        }

        return count;
    }
};
